RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc

```diff
@@ -129,25 +161,69 @@ bool BeamSpotOnlineProducer::shouldShout(const edm::Event& iEvent) const {
   return true;  // Default to "shout" if the record is missing
 }
 
-bool BeamSpotOnlineProducer::processTransientRecord(const edm::EventSetup& iSetup,
-                                                    reco::BeamSpot& result,
-                                                    bool shoutMODE) {
-  auto const& spotDB = iSetup.getData(beamTransientToken_);
-
-  if (spotDB.beamType() != 2) {
-    if (shoutMODE && beamTransientRcdESWatcher_.check(iSetup)) {
-      edm::LogWarning("BeamSpotOnlineProducer")
-          << "Online Beam Spot producer falls back to DB value due to fake beamspot in transient record.";
+bool BeamSpotOnlineProducer::processRecords(const edm::LuminosityBlock& iLumi,
+                                            const edm::EventSetup& iSetup,
+                                            bool shoutMODE) {
+  auto const& spotDBLegacy = getBeamSpotFromRecord(beamTokenLegacy_, iLumi, iSetup, "BeamSpotOnlineLegacyObjectsRcd");
+  auto const& spotDBHLT = getBeamSpotFromRecord(beamTokenHLT_, iLumi, iSetup, "BeamSpotOnlineHLTObjectsRcd");
+  auto const& spotDB = chooseBS(spotDBLegacy, spotDBHLT);
+
+  if (spotDB.beamType() != reco::BeamSpot::Tracker) {
+    if (shoutMODE && (beamLegacyRcdESWatcher_.check(iSetup) || beamHLTRcdESWatcher_.check(iSetup))) {
+      edm::LogWarning("BeamSpotOnlineProducer") << "None of the online records holds a valid beam spot. The Online "
+                                                   "Beam Spot producer falls back to the PCL value.";
     }
     return true;  // Trigger fallback to DB
   }
 
   // Create BeamSpot from transient record
-  createBeamSpotFromTransient(spotDB, result);
+  createBeamSpotFromRecord(spotDB);
   return false;  // No fallback needed
 }
 
-void BeamSpotOnlineProducer::createBeamSpotFromTransient(const BeamSpotObjects& spotDB, reco::BeamSpot& result) const {
+template <typename TokenType>
+const BeamSpotOnlineObjects& BeamSpotOnlineProducer::getBeamSpotFromRecord(const TokenType& token,
+                                                                           const LuminosityBlock& lumi,
+                                                                           const EventSetup& setup,
+                                                                           const std::string whichrecord) {
+  const auto& bs = setup.getData(token);
+  if (bs.sigmaZ() < sigmaZThreshold_ || bs.beamWidthX() < sigmaXYThreshold_ || bs.beamWidthY() < sigmaXYThreshold_ ||
+      bs.beamType() != reco::BeamSpot::Tracker) {
+    edm::LogWarning("BeamSpotOnlineProducer")
+        << "The beam spot record does not pass the fit sanity checks (record: " << whichrecord << ")" << std::endl
+        << "sigmaZ: " << bs.sigmaZ() << std::endl
+        << "sigmaX: " << bs.beamWidthX() << std::endl
+        << "sigmaY: " << bs.beamWidthY() << std::endl
+        << "type: " << bs.beamType();
+    return fakeBS_;
+  }
+  auto lumitime = std::chrono::seconds(lumi.beginTime().value() >> 32);
```

Comment:
How about
```suggestion
  auto lumitime = std::chrono::seconds(lumi.beginTime().unixTime());
```
?

Reply:
@lguzzi before restarting tests, can you address this?